### DISCUSSION TITLE
[Question] Is there any API for navigating layouts?

### DISCUSSION THREAD

**Discussion by roshni-truax** [MunifTanjim/nui.nvim#289] [Q&A]
Hi, 

Thank you for your plugin! I'm making a tool which initially launches three pop-up windows below the cursor. I'd like the user to start in the first one and navigate between the three with `<Tab>`. 

For a number of reasons, this seems to require a surprising amount of work. 

- Firstly, handling the usual task of quitting the layout when leaving is much more complex than in the case of a single popup (because, if one just copies/pastes the usual "on BufLeave" autocmd syntax in the wiki, it'll unmount when moving from one window to the next). 
- Secondly, it doesn't seem like there's a way to refocus onto a new element of the layout. I understand that I can create a hack to do this by having a function which returns a layout which focuses on pop-up **n** on mount, and then unmounting/remounting. Yet this erases the data, so I'd have to store it somewhere. Is there an method I missed for fixing this?
- Finally, I can't use any `<Ctrl-W>` type command to navigate them for three reasons: firstly, none of them work in the general case of multiple windows, secondly, they conflict with the `BufLeave` event, and thirdly, despite setting the third window to be unfocusable, by using `<Ctrl-W>#w` commands for an appropriate number `#`, I am able to focus it using keystrokes.

It seems like this would be a common task, so I'm wondering if I'm missing something with my strategy (e.g. maybe I should be using a Table, or some collection of methods for this task), or if this is a feature gap.

**Comment by MunifTanjim** [comment]
> Firstly, handling the usual task of quitting the layout when leaving is much more complex than in the case of a single popup (because, if one just copies/pastes the usual "on BufLeave" autocmd syntax in the wiki, it'll unmount when moving from one window to the next).

```lua
local Layout = require("nui.layout")
local Popup = require("nui.popup")

local popup = {
  a = Popup({ border = "single" }),
  b = Popup({ border = "single" }),
  c = Popup({ border = "single" }),
}

local layout = Layout(
  {
    relative = "editor",
    position = "50%",
    size = { height = 10, width = 60 },
  },
  Layout.Box({
    Layout.Box(popup.a, { grow = 1 }),
    Layout.Box(popup.b, { grow = 1 }),
    Layout.Box(popup.c, { grow = 1 }),
  }, { dir = "row" })
)

for _, p in pairs(popup) do
  p:on("BufLeave", function()
    vim.schedule(function()
      local bufnr = vim.api.nvim_get_current_buf()
      for _, lp in pairs(popup) do
        if lp.bufnr == bufnr then
          return
        end
      end
      layout:unmount()
    end)
  end)
end

layout:mount()
```

**Reply by MunifTanjim** [reply]
>  I'd like the user to start in the first one

Either do

```lua
vim.api.nvim_set_current_win(popup.a.winid)
```

after `layout:mount()`.

Or set `enter = true` option for `popup.a`.

**Reply by MunifTanjim** [reply]
Added this guide in https://github.com/MunifTanjim/nui.nvim/wiki/nui.layout

**Comment by MunifTanjim** [comment]
> Secondly, it doesn't seem like there's a way to refocus onto a new element of the layout. I understand that I can create a hack to do this by having a function which returns a layout which focuses on pop-up n on mount, and then unmounting/remounting. Yet this erases the data, so I'd have to store it somewhere. Is there an method I missed for fixing this?

I don't think I understand the problem. What exactly are you trying to do? 🤔

---

If you just want to focus a specific popup, you can use `vim.api.nvim_set_current_win` for that.

The window id for a popup is stored in `.winid property. So it'll be something like:

```lua
vim.api.nvim_set_current_win(popup.a.winid)
```

**Reply by roshni-truax** [reply]
Your above comments helped me figure this out: 
 - The revised BufLeave command fixed the issue about moving windows quitting the layout.
 - `vim.api.nvim_set_current_win(popup.a.winid)` is what I needed: I wanted to set a keybind to move from one window to the next and this works.

Thank you!

**Reply by MunifTanjim** [reply]
Added the popup focus guide in https://github.com/MunifTanjim/nui.nvim/wiki/nui.popup

**Comment by MunifTanjim** [comment]
> Finally, I can't use any <Ctrl-W> type command to navigate them for three reasons:

> firstly, none of them work in the general case of multiple windows,

Not sure what you mean.

> secondly, they conflict with the BufLeave event,

I think that would be solved by the suggestion in https://github.com/MunifTanjim/nui.nvim/discussions/289#discussioncomment-6255946 ?

> and thirdly, despite setting the third window to be unfocusable, by using `<Ctrl-W>#w` commands for an appropriate number #, I am able to focus it using keystrokes.

~Again, not sure what you mean. Can you explain a bit more?~

If you're talking about `CTRL-W_w`, then it might be a bug in https://github.com/neovim/neovim. It's letting you focus an unfocusable window using `CTRL-W_w`.

It might also be an expected behavior, because the doc says:

> Non-focusable windows can be entered by |nvim_set_current_win()|.

It might also be true for `CTRL-W_w` with count.

It's not something `nui.nvim` is controlling.

**Reply by roshni-truax** [reply]
Yes, your suggestion fixes that issue. 
And yes, that's the bug that I'm talking about. Thanks for clarifying that it's a neovim issue rather than a nui issue. 

Appreciate all the help! Thanks for your plugin.

**Reply by MunifTanjim** [reply]
Glad I could help 😃

Let me know if you face any more problems.
